lib/wadler/print.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+# Oppen.
+module Oppen
+  # Wadler.
+  class Wadler
+    attr_reader :current_nest_indentation
+    attr_reader :margin
+    attr_reader :new_line
+    attr_reader :tokens
+
+    def initialize(margin: 80, new_line: "\n")
+      @current_nest_indentation = 0
+      @margin = margin
+      @new_line = new_line
+      @tokens = []
+    end
+
+    # @return [String]
+    def output
+      tokens << Oppen.eof
+      Oppen.print(tokens:, margin:, new_line:)
+    end
+
+    # @param indent [Integer] group indentation
+    # @param open_obj [String] group oppening delimiter
+    # @param close_obj [String] group closing delimiter
+    # @param break_type [Oppen::Token::BreakType] group breaking type
+    #
+    # @return [Nil]
+    def group(indent = 0, open_obj = '', close_obj = '',
+              break_type = Oppen::Token::BreakType::CONSISTENT, &)
+      if !open_obj.empty?
+        self.break
+        text(open_obj)
+      end
+
+      nest(indent, break_type, &)
+
+      if !close_obj.empty? # rubocop:disable Style/GuardClause
+        self.break
+        text(close_obj)
+      end
+    end
+
+    # @param indent [Integer] nest indentation
+    # @param break_type [Oppen::Token::BreakType] nest breaking type
+    #
+    # @return [Nil]
+    def nest(indent = 2, break_type = Oppen::Token::BreakType::CONSISTENT, &block)
+      last_nest_indentation = current_nest_indentation
+      @current_nest_indentation = indent
```

Comment:
How about we simplify the names and call them `previous_indent`, and `current_indent`?